### DISCUSSION TITLE
How hard would it be to enable python indentation for V?

### DISCUSSION THREAD

**Discussion by jokoon** [vlang/v#16843] [Ideas]
I'm currently exploring and trying to make a language quite similar to V (light in features, string/dict/list/tuple, translates to C), except I want python indentation.

Have you already explored this? I'm not really sure, but it seems it would not affect V's syntax very much. Indent-parsing only happens at the lexing phase.

**Comment by JalonSolov** [comment]
It doesn't matter how much it would affect things, it is extremely unlikely to ever happen in V.  Good luck in your own project.

**Reply by jokoon** [reply]
DId you write the grammar of V in any grammar like BNF, yacc or other similar language?

**Reply by JalonSolov** [reply]
I didn't write it, but no, it is custom, hand-written code.

One of the reasons this change is unlikely to happen in V is that all whitespace is basically ignored.

In V,
```v
fn main() {
    for _ in 0..10 {
        println('hello')
    }
}
```
and
```v
fn main(){for _ in 0..10{println('hello')}}
```
are treated identically.  Whitespace, including linefeeds, is basically just a separator (and no matter how much, it's treated as a single space), and is only required when necessary (such as between `fn` and `main`)

The formatter, `v fmt` formats the code consistently (with tab indents instead of spaces), but that is to make it more readable, not for the compiler to parse it.

It would be a complete 180 degree turn to change to indent-based.

**Comment by jokoon** [comment]
Indenting could be done before parsing and add curly braces, I guess it
could be some optional compiling flag, maybe at the start of a V file, but
of course I did not mean to change V and make it indent based, I was just
asking because I'm not very experienced in parsing, and so I was not sure
it was possible or easy.

But I understand if you don't want to, it might hurt the compiling
performance and it might also introduce some confusion for users.

I was asking for the V grammar because the C grammar can be a bit weird
(direct-abstract-declarator parses function signatures), and I wanted to
see how V is parsed.

**Reply by JalonSolov** [reply]
The only way to see how V is parsed is to read the compiler code.

**Comment by Wajinn** [comment]
> Indenting could be done before parsing and add curly braces, I guess it could be some optional compiling flag, maybe at the start of a V file, but of course I did not mean to change V and make it indent based, I was just asking because I'm not very experienced in parsing, and so I was not sure it was possible or easy.

V, as various other languages, are not whitespace/off-side rule based. Arguably, most/many do not like Python-like indentation rules. Additionally, whitespace indentation/off-side rules come with very significant problems. Of course there are those coming from such languages that may have become accustomed or have a preference for (whitespace/off-side), but attempting to modify the design of a different language, with different long established preferences, is a bit much.

It would be like V users coming to Python forums and asking them to modify their language to accept curly braces and tabs. Another, would be C users coming to Pascal forums asking them to get rid of begin...end and use {...}. Those incursions don't usually go well. Most users of the language would probably want the developers to focus on other things instead. And actually it is worse for Python fans to attempt to derail V (a developing language in beta), then the other way around.

Another point is V positions itself as a C-like language, so the request is not so minor as it might seem, but a major departure from the design and established philosophy.

But, I kind of do think that if an individual or group does want whitespace/off-side, they possibly could make a module for the language, which could act as a preprocessor or light-transpiler, and see how that goes. For instance, this has been done at various times for Python. For those with the opposite preference that like Python, but prefer C-like curly brackets and tabs.

**Comment by JoanaBLate** [comment]
Hi, once I started creating a language (with Python indentation) that translates to JS. I changed the project into a JavaScript linter that I use every day (the code is in my GitHub account). 

Before giving up implementing the language, I wrote the code that handles the meaningful indentation. It is easy. I did this way (*simplifying*):

1) Create a list of tokens from the source code file.

2) Neighbor white spaces must be grouped in the same token.

3) White space tokens that start the line are converted to margin tokens (always multiple of 4 (you may choose other length) white spaces or you have a syntax error), the other ones you discharge: they already did their job (which is separating content tokens).

4) While you parse token by token, each time you find a margin token, you compare it with the previous margin token you got. 

4.1) If they have the same length, nothing happens. 

4.2) If the recent token length is bigger, this means a code block is being opened, it is like you got the token "{" in V. You must check if it is a good moment to open a code block or if it is a syntax error in the source file. Also, the difference between the margins cannot be greater than 4 white spaces (1 indent).

4.3) If the recent token length is smaller, this means one or more code blocks are being closed. It is like you got one or more tokens "}" in V. You must check if it is a good moment to close that/those blocks or if it is a syntax error in the source file.

This is the logic that I used and worked perfectly. It complicates a bit when you have one logic line using more than one physical line. You have to decide how the language handles line continuations (same indentation? bigger? smaller?) and consider in your parser that that is a line continuation and not a real change in indentation (opening or closing block).

I am not proposing changing V, it is just to help you with your own language.

**Reply by Wajinn** [reply]
I get what you are saying. Seems like creating a Vlang module for Python-like whitespace/off-side rules, which is a preprocessor/light-transpiler of Python-like rules to standard Vlang code, seems like the way to go.

**Reply by JoanaBLate** [reply]
Yes. You can easily preprocess a Python like file inserting the correct curly braces (sorry, Guido van Rossum). 

But I used a more direct technique: for each single difference in indentation, I inserted an artificial token in the tokens list: an open-block token or close-block token.

I mean, I did not have to preprocess the source file.

Again, I am not proposing to change V.

**Comment by bogen85** [comment]
Even if someone wanted to and had the time and resources to do this, even as a separate module, it would not be possible with the current V syntax.
(unless whoever was wanting to do this was willing to support a subset of the language).


Take this example from here https://github.com/vlang/v/blob/master/doc/docs.md#channel-select : 
```v
import time

fn main() {
	ch := chan f64{}
	ch2 := chan f64{}
	ch3 := chan f64{}
	mut b := 0.0
	c := 1.0
	// ... setup spawn threads that will send on ch/ch2
	spawn fn (the_channel chan f64) {
		time.sleep(5 * time.millisecond)
		the_channel <- 1.0
	}(ch)
	spawn fn (the_channel chan f64) {
		time.sleep(1 * time.millisecond)
		the_channel <- 1.0
	}(ch2)
	spawn fn (the_channel chan f64) {
		_ := <-the_channel
	}(ch3)

	select {
		a := <-ch {
			// do something with `a`
			eprintln('> a: ${a}')
		}
		b = <-ch2 {
			// do something with predeclared variable `b`
			eprintln('> b: ${b}')
		}
		ch3 <- c {
			// do something if `c` was sent
			time.sleep(5 * time.millisecond)
			eprintln('> c: ${c} was send on channel ch3')
		}
		500 * time.millisecond {
			// do something if no channel has become ready within 0.5s
			eprintln('> more than 0.5s passed without a channel being ready')
		}
	}
	eprintln('> done')
}
```

The above does not lend itself to side rule.

```v
	spawn fn (the_channel chan f64) {
		time.sleep(5 * time.millisecond)
		the_channel <- 1.0
	}(ch)
```

There is no structure in python like that.
A the dangling (ch) on a line of it's own makes more sense.
Either:
1. The use of channels in that way would not be supported.
2. One would come up with an alternate syntax for that where that is nothing following the closing brace.

It in unlikely that such a syntax change would be allowed in core V just o accommodate such a module.
Without such a syntax change, who would want a module like this that only supports a subset of the language?

**Reply by JoanaBLate** [reply]
I am in favor of keeping the current V syntax. That said, the algorithm for opening/closing blocks works fine for your example too. The visible result is ugly and lacks clarity, but the algorithm works: 
```
	spawn fn (the_channel chan f64) 
		time.sleep(5 * time.millisecond)
		the_channel <- 1.0
	(ch)
```

**Reply by MatejMagat305** [reply]
I would vote for current syntax too ...

**Reply by bogen85** [reply]
As @JalonSolov pointed out, this would in all likelihood need to be project outside of the core github vlang/v project.
Honestly I don't see such a module being accepted, ever. V evolved mostly from Go, not from Python.
Even if such a module was was proven to work well, I don't how it would be accepted.

Without it being accepted, how would someone use it in a work flow? In a seamless manner? A source tree is multiple files, how would it know which to convert and where to put them?

Would this be a wrapper for V if not accepted into the core V?

**Reply by JalonSolov** [reply]
Either a wrapper or a preprocessor.  Convert the python-ish source to regular V, then call V to compile it.

I agree that this will likely never be part of V itself.

**Reply by bogen85** [reply]
Since this is a separate language project, I guess not an issue on any of my previous workflow questions.

**Reply by Dairoe** [reply]
is it really that hard though? 

this code here

which v explains is like c++,c style loop

`for i=0; i <0; i++ {}`

look at the ';' this out of the whole v lang i seen this used how would it be hard to tokenize and indent if we have a loop thats from old style

**Reply by Dairoe** [reply]
rust supports tabs and spaces but recmd spaces, same goes for other lang.  saying its not possible when its a file with call v from python file inside the example, but yes changing the lang from {} to indent based it out of question and should be sep
